webapp/crashstats/tests/test_libbugzilla.py

```diff
@@ -0,0 +1,677 @@
+        """Verify moz_crash_reason shows up but not reason when they're both there."""
+        processed_crash = {
+            "reason": "EXCEPTION_BREAKPOINT",
+            "moz_crash_reason": "MOZ_CRASH(Quota manager shutdown timed out) (good)",
+            "moz_crash_reason_raw": "MOZ_CRASH(Quota manager shutdown timed out) (bad)",
+            "crashing_thread": 0,
+            "json_dump": {
+                "threads": [
+                    {
+                        "frames": [
+                            {
+                                "frame": 0,
+                                "module": "fake_module",
+                                "signature": "foo::bar(char* x, int y)",
+                                "file": "fake.cpp",
+                                "line": 1,
+                            },
+                        ],
+                    }
+                ],
+            },
+        }
+        comment = crash_report_to_description(self.URL, processed_crash)
+        assert comment == dedent(
+            """\
+            Crash report: http://localhost:8000/report/index/2ae0a833-f43d-4d9b-8c13-f99e70240401
+
+            MOZ_CRASH Reason: ```MOZ_CRASH(Quota manager shutdown timed out) (good)```
+
+            Top 1 frame:
+            ```
+            0  fake_module  foo::bar(char* x, int y)  fake.cpp:1
+            ```"""
+        )
+
+    def test_comment_crashing_thread_none(self):
+        """Verify no crashing thread is treated as thread 0 with note."""
+        processed_crash = {
+            "json_dump": {
+                "threads": [
+                    {
+                        "frames": [
+                            {
+                                "frame": 0,
+                                "module": "fake_module",
+                                "signature": "foo::bar(char* x, int y)",
+                                "file": "fake.cpp",
+                                "line": 1,
+                            },
+                        ],
+                    }
+                ],
+            },
+        }
+        comment = crash_report_to_description(self.URL, processed_crash)
+        assert comment == dedent(
+            """\
+            Crash report: http://localhost:8000/report/index/2ae0a833-f43d-4d9b-8c13-f99e70240401
+
+            No crashing thread identified; using thread 0.
+
+            Top 1 frame:
+            ```
+            0  fake_module  foo::bar(char* x, int y)  fake.cpp:1
+            ```"""
+        )
+
+    def test_comment_java_stack_trace(self):
+        """If there's a java stack trace, use that instead
+
+        Also verify tabs get converted to 4-spaces.
+
+        """
+        processed_crash = {
+            "java_stack_trace": dedent(
+                """\
+                java.lang.NoClassDefFoundError: kotlinx.coroutines.channels.BufferedChannel
+                \tat kotlinx.coroutines.channels.ChannelKt.Channel$default(Channel.kt:44)
```

Comment:
The `\t` inserts a tab. The `java_stack_trace` values have tabs at the beginning of stack frame lines. Fun times.